File: stone_lib/src/main/java/com/github/klee0kai/stone/_hidden_/provide/ProvideBuilder.java
package com.github.klee0kai.stone._hidden_.provide;

import java.util.List;

public class ProvideBuilder<T> {

    public interface ProvideBody<T> {
        void provide(ProvideConsumer<T> consumer);
    }

    private ProvideBody<T> provideBody;

    public ProvideBuilder(ProvideBody<T> provideBody) {
        this.provideBody = provideBody;
    }

    public T first() {
        ProvideConsumer<T> consumer = new ProvideConsumer<>();
        provideBody.provide(consumer);
        return consumer.getFirst();
    }

    public List<T> all() {
        ProvideConsumer<T> consumer = new ProvideConsumer<>();
        provideBody.provide(consumer);
        return consumer.getList();
    }
}

File: stone_multiplatform/src/commonMain/kotlin/com/github/klee0kai/stone/__hidden__/provide/ProvideBuilder.kt
package com.github.klee0kai.stone.__hidden__.provide

class ProvideBuilder<T>(
    private val provideBody: ProvideBody<T>
) {

    fun interface ProvideBody<T> {
        fun provide(consumer: ProvideConsumer<T>)
    }

    fun firstOrNull(): T? {
        val consumer = ProvideConsumer<T>()
        provideBody.provide(consumer)
        return consumer.first
    }


    fun first(): T {
        val consumer = ProvideConsumer<T>()
        provideBody.provide(consumer)
        return consumer.first!!
    }

    fun all(): List<T> {
        val consumer = ProvideConsumer<T>()
        provideBody.provide(consumer)
        return consumer.list
    }

}


